include/grpcpp/impl/codegen/async_unary_call.h

```diff
@@ -66,26 +72,133 @@ class ClientAsyncResponseReaderInterface {
+  /// (e.g., MessageLite for protobuf). Since constructors can't have an
+  /// explicit template parameter, the last argument is a bogus to the
+  /// constructor just to provide the needed type information.
+  template <class R, class W, class BaseR = R, class BaseW = W>
+  static ClientAsyncResponseReader<R>* Create(
+      ::grpc::ChannelInterface* channel, ::grpc::CompletionQueue* cq,
+      const ::grpc::internal::RpcMethod& method, ::grpc::ClientContext* context,
+      const W& request) /* __attribute__((noinline)) */ {
+    ::grpc::internal::Call call = channel->CreateCall(method, context, cq);
+    ClientAsyncResponseReader<R>* result =
+        new (::grpc::g_core_codegen_interface->grpc_call_arena_alloc(
+            call.call(), sizeof(ClientAsyncResponseReader<R>)))
+            ClientAsyncResponseReader<R>(call, context);
+    SetupRequest<BaseR, BaseW>(
+        call.call(), &result->single_buf_, &result->read_initial_metadata_,
+        &result->finish_, static_cast<const BaseW&>(request));
+
+    return result;
+  }
+
+  // Various helper functions to reduce templating use
+
+  template <class R, class W>
+  static void SetupRequest(
+      grpc_call* call,
+      ::grpc::internal::CallOpSendInitialMetadata** single_buf_ptr,
+      std::function<void(ClientContext*, internal::Call*,
+                         internal::CallOpSendInitialMetadata*, void*)>*
+          read_initial_metadata,
+      std::function<
+          void(ClientContext*, internal::Call*, bool initial_metadata_read,
+               internal::CallOpSendInitialMetadata*,
+               internal::CallOpSetInterface**, void*, Status*, void*)>* finish,
+
+      const W& request) {
+    using SingleBufType =
+        ::grpc::internal::CallOpSet<::grpc::internal::CallOpSendInitialMetadata,
+                                    ::grpc::internal::CallOpSendMessage,
+                                    ::grpc::internal::CallOpClientSendClose,
+                                    ::grpc::internal::CallOpRecvInitialMetadata,
+                                    ::grpc::internal::CallOpRecvMessage<R>,
+                                    ::grpc::internal::CallOpClientRecvStatus>;
+    SingleBufType* single_buf =
+        new (::grpc::g_core_codegen_interface->grpc_call_arena_alloc(
+            call, sizeof(SingleBufType))) SingleBufType;
+    *single_buf_ptr = single_buf;
+    // TODO(ctiller): don't assert
+    GPR_CODEGEN_ASSERT(single_buf->SendMessage(request).ok());
+    single_buf->ClientSendClose();
+
+    *read_initial_metadata =
+        [](ClientContext* context, internal::Call* call,
+           internal::CallOpSendInitialMetadata* single_buf_view, void* tag) {
+          auto* single_buf = static_cast<SingleBufType*>(single_buf_view);
+          single_buf->set_output_tag(tag);
+          single_buf->RecvInitialMetadata(context);
+          call->PerformOps(single_buf);
+        };
+
+    *finish = [](ClientContext* context, internal::Call* call,
```

Comment:
Would appreciate a comment on why you're doing these tricks, at least in the first place you do them. :)